nixos/modules/services/networking/crab-hole.nix

```diff
@@ -0,0 +1,352 @@
+                lists = lists.forEach cfg.config.blocklist.lists (
+                  v: if isPath v then ("file://${cfg.workDir}/${filename v}") else v
+                );
+                allow_list = lists.forEach cfg.config.blocklist.allow_list (
+                  v: if isPath v then ("file://${cfg.workDir}/${filename v}") else v
+                );
+              };
+              downstream = forEach cfg.config.downstream (
+                x:
+                if x.protocol == "udp" then
+                  {
+                    protocol = x.protocol;
+                    listen = x.listen;
+                    port = x.port;
+                  }
+                else if (x.protocol == "https" && x.dns_hostname != null) then
+                  x
+                else
+                  {
+                    protocol = x.protocol;
+                    listen = x.listen;
+                    port = x.port;
+                    certificate = x.certificate;
+                    key = x.key;
+                    timeout = x.timeout;
+                  }
+              );
+              upstream.name_servers = cfg.config.upstream.name_servers;
+            }
+            // (
+              if (cfg.config.api == null) then
+                { }
+              else
+                {
+                  api =
+                    if (cfg.config.api.admin_key != null) then
+                      cfg.config.api
+                    else
+                      {
+                        port = cfg.config.api.port;
+                        listen = cfg.config.api.listen;
+                        show_doc = cfg.config.api.show_doc;
+                      };
+                }
+            );
+
+          blockListFiles = builtins.filter (v: isPath v) cfg.config.blocklist.lists;
+          allowListFiles = builtins.filter (v: isPath v) cfg.config.blocklist.allow_list;
+
+          selectedConfig =
+            if cfg.configFile != null then cfg.configFile else (configFormat.generate "config.toml" settings);
+        in
+        ''
+          cp -f '${selectedConfig}' '${cfg.workDir}/config.toml'
+          ${builtins.concatStringsSep "\n" (
+            map (file: "cp -f '${file}' '${cfg.workDir}/${filename file}'") blockListFiles
+          )}
+          ${builtins.concatStringsSep "\n" (
+            map (file: "cp -f '${file}' '${cfg.workDir}/${filename file}'") allowListFiles
+          )}
+        '';
+    };
+
+    systemd.tmpfiles.rules = [
+      "d '${cfg.workDir}' 0750 ${cfg.user} ${cfg.group} - -"
+    ];
+
+    # Adding crab-hole user and group
+    users.users = optionalAttrs (cfg.user == "crab-hole") {
+      crab-hole = {
+        description = "Crab-hole service";
+        home = cfg.workDir;
+        group = cfg.group;
+        isSystemUser = true;
+      };
+    };
+
+    users.groups = optionalAttrs (cfg.group == "crab-hole") {
+      crab-hole = { };
+    };
```

Comment:
This can be replaced by [`DynamicUser`](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#DynamicUser=)